src/Razor/test/Microsoft.VisualStudio.LanguageServerClient.Razor.Test/LSPRazorProjectHostTest.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Moq;
+using Xunit;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    public class LSPRazorProjectHostTest
+    {
+        [Fact]
+        public async Task LoadAsync_NoopsWhenLSPEditorFeatureNotAvailable()
+        {
+            // Arrange
+            var featureDetector = Mock.Of<LSPEditorFeatureDetector>(f => f.IsLSPEditorFeatureEnabled() == false);
+            var broker = new Lazy<ILanguageClientBroker>();
+            var client = new Lazy<ILanguageClient, IDictionary<string, object>>(
+                () => throw new NotImplementedException(), new Dictionary<string, object>());
+            var languageClients = new[] { client };
+
+            var host = new LSPRazorProjectHost(featureDetector, broker, languageClients);
+
+            // Act & Assert (does not throw)
+            await host.LoadAsync().ConfigureAwait(false);
+        }
+
+        [Fact]
+        public async Task LoadAsync_LSPEditorFeatureAvailable_LoadsClients()
+        {
+            // Arrange
+            var featureDetector = Mock.Of<LSPEditorFeatureDetector>(f => f.IsLSPEditorFeatureEnabled() == true);
+
+            var metadata = new Dictionary<string, object>()
+            {
+                { "ClientName", "RazorClient" },
+                { "ContentTypes", new[] { "RazorLSP" } }
+            };
+            var client = new Lazy<ILanguageClient, IDictionary<string, object>>(() => Mock.Of<ILanguageClient>(), metadata);
+            var languageClients = new[] { client };
+
+            var loaded = false;
+            var brokerMock = new Mock<ILanguageClientBroker>();
+            brokerMock
+                .Setup(b => b.LoadAsync(It.IsAny<ILanguageClientMetadata>(), It.IsAny<ILanguageClient>()))
+                .Returns(Task.CompletedTask)
+                .Callback<ILanguageClientMetadata, ILanguageClient>((metadata, c) =>
```

Comment:
Interesting pattern. So after returning the value of `LoadAsync` this callback will be run against the parameters that were passed to it? Whats the advantage of hosting this here as opposed to inside the returns function?

Reply:
The callback here is used to verify if the method was called with the correct parameters. We will not be able to assert the parameters in the `Returns` method because all it takes is the result that it should provide to the original caller.

Reply:
I feel like the more idiomatic way to do that would be something like:
```c#
brokerMock.Setup(b => b.LoadAsync(It.Is<ILanguageClientMetadata>(m => m.ContentTypes.Count == 1 && m.ContentTypes[0] == "RazorLSP"), It.Is<ILanguageClient>(c => c == client.Value));

brokerMock.VerifyAll();
```

Yours DOES have the benefit of producing more detailed error messages, but IMO it's harder to tell that the intention is to verify the parameters passed to the Setup. @NTaylorMullen do you have opinions on which way is better (or a third way)?

Reply:
I agree that feels a bit more natural and intuitive but the `Setup` gets easily cluttered if the number of parameters and the number of things you want to assert increases.

Reply:
I don't have a preference to be honest

Reply:
Sounds like personal preference then, follow your heart.